Source/Particles/Gather/FieldGather.H

```diff
@@ -16,413 +16,305 @@
-        // Gather field on particle Bzp from field on grid bz_arr
-        for (int iz=0; iz<=depos_order; iz++){
-            for (int ix=0; ix<=depos_order-galerkin_interpolation; ix++){
-                const amrex::Real dEx = (+ ex_arr(lo.x+j_ex+ix, lo.y+l_ex+iz, 0, 2*imode-1)*xy.real()
-                                         - ex_arr(lo.x+j_ex+ix, lo.y+l_ex+iz, 0, 2*imode)*xy.imag());
-                Erp += sx_ex[ix]*sz_ex[iz]*dEx;
-                const amrex::Real dBz = (+ bz_arr(lo.x+j_bz+ix, lo.y+l_bz+iz, 0, 2*imode-1)*xy.real()
-                                         - bz_arr(lo.x+j_bz+ix, lo.y+l_bz+iz, 0, 2*imode)*xy.imag());
-                Bzp += sx_bz[ix]*sz_bz[iz]*dBz;
-            }
-        }
-        // Gather field on particle Ezp from field on grid ez_arr
-        // Gather field on particle Brp from field on grid bx_arr
-        for (int iz=0; iz<=depos_order-galerkin_interpolation; iz++){
-            for (int ix=0; ix<=depos_order; ix++){
-                const amrex::Real dEz = (+ ez_arr(lo.x+j_ez+ix, lo.y+l_ez+iz, 0, 2*imode-1)*xy.real()
-                                         - ez_arr(lo.x+j_ez+ix, lo.y+l_ez+iz, 0, 2*imode)*xy.imag());
-                Ezp += sx_ez[ix]*sz_ez[iz]*dEz;
-                const amrex::Real dBx = (+ bx_arr(lo.x+j_bx+ix, lo.y+l_bx+iz, 0, 2*imode-1)*xy.real()
-                                         - bx_arr(lo.x+j_bx+ix, lo.y+l_bx+iz, 0, 2*imode)*xy.imag());
-                Brp += sx_bx[ix]*sz_bx[iz]*dBx;
-            }
-        }
-        // Gather field on particle Bthetap from field on grid by_arr
-        for (int iz=0; iz<=depos_order-galerkin_interpolation; iz++){
-            for (int ix=0; ix<=depos_order-galerkin_interpolation; ix++){
-                const amrex::Real dBy = (+ by_arr(lo.x+j_by+ix, lo.y+l_by+iz, 0, 2*imode-1)*xy.real()
-                                         - by_arr(lo.x+j_by+ix, lo.y+l_by+iz, 0, 2*imode)*xy.imag());
-                Bthetap += sx_by[ix]*sz_by[iz]*dBy;
-            }
-        }
-        xy = xy*xy0;
-    }
+    int jp_Fy = 0;
+    double sy_Fy[depos_order_para + 1] = {0.};
+    if (Fy_type[1] == NODE) { jp_Fy = shape_factor_para(sy_Fy, y_bar); }
+    else  { jp_Fy = shape_factor_para(sy_Fy, y_bar - 0.5_rt); }
 
-    // Convert Erp and Ethetap to Ex and Ey
-    Exp += costheta*Erp - sintheta*Ethetap;
-    Eyp += costheta*Ethetap + sintheta*Erp;
-    Bxp += costheta*Brp - sintheta*Bthetap;
-    Byp += costheta*Bthetap + sintheta*Brp;
-
-#else // defined(WARPX_DIM_3D)
-    // Gather field on particle Exp from field on grid ex_arr
-    for (int iz=0; iz<=depos_order; iz++){
-        for (int iy=0; iy<=depos_order; iy++){
-            for (int ix=0; ix<= depos_order - galerkin_interpolation; ix++){
-                Exp += sx_ex[ix]*sy_ex[iy]*sz_ex[iz]*
-                    ex_arr(lo.x+j_ex+ix, lo.y+k_ex+iy, lo.z+l_ex+iz);
+    // Gather vector field F
+    amrex::Real weight;
+    for (int i=0; i<=depos_order_para; i++) {
```

Comment:
For optimization, the loop ordering should be swapped, with `i` the innermost loop. The same for all of the nested loops.